torch/nn/parallel/distributed.py

```diff
@@ -714,7 +715,20 @@ def _recursive_to(self, inputs, target_gpu):
         """
         def to_map(obj):
             if isinstance(obj, torch.Tensor):
-                return (obj.to(target_gpu), )
+                # Perform CPU -> GPU copies in a background stream. This code is
+                # motivated from similar logic in torch/nn/parallel/_functions.py
+                stream = _get_stream(target_gpu)
+                with torch.cuda.stream(stream):
+                    output = obj.to(target_gpu)
+                # synchronize with the copy stream
+                with torch.cuda.device(target_gpu):
+                    current_stream = torch.cuda.current_stream()
+                    # Sync the current stream with the copy stream
+                    current_stream.wait_stream(stream)
```

Comment:
isn't the copy operator already doing this in c++ code?

https://github.com/pytorch/pytorch/blob/7b21c6be67bc8669e7b93e641a363bbb84b4856d/aten/src/ATen/native/cuda/Copy.cu#L90

Reply:
Good point, although I wanted to keep this change as similar as `scatter_gather` code as possible to prevent regressions/errors: https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/_functions.py#L102. We can probably evaluate whether to change this in both callsites in a follow-up change?